js/csv.js

```diff
@@ -0,0 +1,230 @@
+/**
+ * Reads a CSV file and returns a table with the data.
+ *
+ * @param {Blob|File} file The file object to be read.
+ * @param {Function} parser The function to be used in the
+ * CSV parsing operation.
+ */
+export const readCsv = (file, parser = null) => {
+    parser = parser || parseCsvComplex;
+    return new Promise((resolve, reject) => {
+        const reader = new FileReader();
+        reader.onload = e => resolve(parser(e.target.result));
+        reader.onerror = e => reject(e);
+        reader.readAsText(file);
+    });
+};
+
+/**
+ * Creates the contents of the CSV file with the data given
+ * as an array of rows and the headers.
+ *
+ * @param {Array / Object} data Data to be converted into CSV. Can be
+ * an array of strings, a JSON object, or an array of those.
+ * @param {Array} headers List of headers for the CSV file, is an array of strings.
+ * @param {String} delimiter The delimiter to be used in the CSV.
+ * @returns A string containing the information for the CSV.
+ */
+export const buildCsv = (data, headers = [], delimiter = ",") => {
+    let csv = "";
+    if (Object.keys(headers).length !== 0) csv += headers.join(delimiter) + "\n";
+
+    // is simple object
+    if (!Array.isArray(data)) {
+        return csv + _objectToCsv(data, headers, delimiter);
+    }
+    // is simple array
+    if (typeof data[0] === "string") return csv + _arrayToCsv(data, delimiter);
+    if (Object.keys(data).length === 0) return "";
+
+    for (let i = 0; i < Object.keys(data).length; i++) {
+        // check if array, if it is add value as is
+        if (Array.isArray(data[i])) {
+            csv += _arrayToCsv(data[i], delimiter);
+            continue;
+        }
+
+        csv += _objectToCsv(data[i], headers, delimiter);
+    }

```

Comment:
This would not work for example if it was an array where the first element was an array of arrays. I think instead of a loop here you can try a recursive solution, sort of like https://github.com/ripe-tech/ripe-commons-js/blob/pr/add_csv_functions/js/csv.js#L224

Parse recursively, where if it is something simple, just toString it, if it is an object, do your object to CSV, if it is an array, call recursively for each element.

Reply:
I get what you mean, but what would an array of array of array of strings even look like as a CSV?

Reply:
There could be where the first array is the entire line, and then a field which is an array of arrays. I don't see it being often, but could happen, and a recursive solution will take care of any level of nestedness.

Reply:
Writing a test for this, I don't think it's that necessary to add that level of complexity (at least not yet), especially when it might just result in a strangely formatted CSV. And even in that case, if there's multiple nested lists, how would it look like? Currently, if there's `["a", ["b", ["c", "d"]]`, it returns `a,"b,c,d"`, since it concatenates all values. This would help align the value with the correct header, instead of potentially "spilling over" into other columns. There's already [a test for this](https://github.com/ripe-tech/ripe-commons-js/blob/07ad06e658efb9fc60c93d5734022f3c323bf4b6/test/csv.js#L120), wouldn't this be enough?

Reply:
Sure